File: Animations/Sunrays/Golang/sunray.go
package main

import rl "github.com/gen2brain/raylib-go/raylib"

type SunRay struct {
	mouse     *Mouse
	effect    *Effect
	start     []*rl.Vector2
	end       *rl.Vector2
	color     *rl.Color
	total     int32
	divisor   int32
	thickness float32
	opacity   float32
}

// ---------------------------------------------------------------------------------------------------------------------

func (s *SunRay) init() {
	for i := range s.effect.totalParticles {
		if i%s.divisor == 0 {
			s.total++
		}
	}

	s.color = &rl.Color{R: 255, G: 255, B: 255, A: uint8(255 * s.opacity)}
}

// ---------------------------------------------------------------------------------------------------------------------

func (s *SunRay) calculateLocations() {
	s.start = make([]*rl.Vector2, 0)
	for i := range s.effect.totalParticles {
		if i%s.divisor == 0 {
			s.start = append(
				s.start,
				&rl.Vector2{
					X: float32(s.effect.particles[i].x),
					Y: float32(s.effect.particles[i].y),
				},
			)
		}
	}

	s.end = &rl.Vector2{
		X: float32(s.mouse.x),
		Y: float32(s.mouse.y),
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (s *SunRay) draw() {
	for i := range s.total {
		rl.DrawLineEx(
			*s.start[i],
			*s.end,
			s.thickness,
			*s.color,
		)
	}
}

// ---------------------------------------------------------------------------------------------------------------------

File: Animations/Sunrays/Python/sunray.py
import pyray as rl
from particle import Particle
from mouse import Mouse


class Sunray:
    def __init__(
        self,
        particles: list[Particle],
        mouse: Mouse,
        ray_divisor: int,
        opacity: float,
        thickness: float,
    ) -> None:
        self.particles = particles
        self.start_x: list[rl.Vector2] = []
        self.mouse = mouse
        self.ray_divisor = ray_divisor
        self.thickness = thickness
        self.color = rl.Color(255, 255, 255, int(255 * opacity))

    def calculate_rays(self) -> None:
        self.start_x = []
        for i, particle in enumerate(self.particles):
            if i % self.ray_divisor == 0:
                self.start_x.append(rl.Vector2(particle.x, particle.y))

    def draw_rays(self) -> None:
        endpoint = rl.Vector2(self.mouse.x, self.mouse.y)

        for ray_start in self.start_x:
            rl.draw_line_ex(
                ray_start,
                endpoint,
                self.thickness,
                self.color,
            )
